File: clients/csharp/Runner.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using System.Net.Sockets;
using System.Text;

namespace AiCup2019
{
    public class Runner
    {
        private BinaryReader reader;
        private BinaryWriter writer;
        public Runner(string host, int port, string token)
        {
            var client = new TcpClient(host, port) { NoDelay = true };
            var stream = new BufferedStream(client.GetStream());
            reader = new BinaryReader(stream);
            writer = new BinaryWriter(stream);
            var tokenData = System.Text.Encoding.UTF8.GetBytes(token);
            writer.Write(tokenData.Length);
            writer.Write(tokenData);
            writer.Flush();
        }
        public void Run()
        {
            var myStrategy = new MyStrategy();
            var debug = new Debug(writer);
            while (true)
            {
                Model.ServerMessageGame message = Model.ServerMessageGame.ReadFrom(reader);
                if (!message.PlayerView.HasValue)
                {
                    break;
                }
                Model.PlayerView playerView = message.PlayerView.Value;
                var actions = new Dictionary<int, Model.UnitAction>();
                foreach (var unit in playerView.Game.Units)
                {
                    if (unit.PlayerId == playerView.MyId)
                    {
                        actions.Add(unit.Id, myStrategy.GetAction(unit, playerView.Game, debug));
                    }
                }
                new Model.PlayerMessageGame.ActionMessage(new Model.Versioned(actions)).WriteTo(writer);
                writer.Flush();
            }
        }
        public static void Main(string[] args)
        {
            string host = args.Length < 1 ? "127.0.0.1" : args[0];
            int port = args.Length < 2 ? 31001 : int.Parse(args[1]);
            string token = args.Length < 3 ? "0000000000000000" : args[2];
            new Runner(host, port, token).Run();
        }
    }
}

File: clients/fsharp/Runner.fs
namespace AiCup2019

open System
open System.IO;
open System.Net.Sockets;

module Runner =
    type T(host, port, token: string) =
        let client = new TcpClient(host, port)
        let stream = new BufferedStream(client.GetStream())
        let reader = new BinaryReader(stream)
        let writer = new BinaryWriter(stream)
        let tokenData = System.Text.Encoding.UTF8.GetBytes token
        do 
            client.NoDelay <- true
            writer.Write tokenData.Length
            writer.Write tokenData
            writer.Flush()

        member this.run =
            let myStrategy = new MyStrategy()
            let debug = new Debug(writer)

            let rec loop() = 
                let message = Model.ServerMessageGame.readFrom reader
                
                match message.PlayerView with
                    | Some playerView ->
                        let actions = playerView.Game.Units 
                                                |> Array.filter(fun x -> x.PlayerId = playerView.MyId)
                                                |> Array.map(fun x -> 
                                                    (x.Id, myStrategy.getAction(x, playerView.Game, debug))) 
                                                |> Map.ofArray                                                       
                        (Model.PlayerMessageGame.ActionMessage {Action = {Inner = actions}}).writeTo writer
                        writer.Flush()
                        loop()
                    | None -> ()

            loop()

    [<EntryPoint>]
    let main argv =
        let host = match argv with
                    | x when x.Length >=1 -> x.[0]
                    | _-> "127.0.0.1"
        let port = match argv with
                    | x when x.Length >=2 -> x.[1] |> Int32.Parse
                    | _ -> 31001
        let token = match argv with
                    | x when x.Length >=3 -> x.[2]
                    | _ -> "0000000000000000"

        (new T(host, port, token)).run

        0 // return an integer exit code
